### ISSUE TITLE
Installation issue: mpfr

### ISSUE THREAD

**Issue by darmac** [LLNL/spack#16140]
mpfr fetch patch failed

### Steps to reproduce the issue
```console
$ spack install -v mpfr
[/home/xiaojun/spack/var/spack/repos/builtin/packages]$siv mpfr
[+] /home/xiaojun/spack/opt/spack/linux-centos8-aarch64/gcc-8.2.1/libsigsegv-2.12-psetehw3pobuucu6vvaao5b3rkrbehzn
[+] /home/xiaojun/spack/opt/spack/linux-centos8-aarch64/gcc-8.2.1/pkgconf-1.6.3-xwwcou2uvone45j36scf7atvesbadmah
[+] /home/xiaojun/spack/opt/spack/linux-centos8-aarch64/gcc-8.2.1/m4-1.4.18-5hu7gsgnj4sunqjnhix74m7gh76uh4at
[+] /home/xiaojun/spack/opt/spack/linux-centos8-aarch64/gcc-8.2.1/ncurses-6.2-5zs46oqspomfxll4jasxilx3xnw24ylc
[+] /home/xiaojun/spack/opt/spack/linux-centos8-aarch64/gcc-8.2.1/libtool-2.4.6-2hqoi4hxaxurgvdxamcio5v4cojdztpw
[+] /home/xiaojun/spack/opt/spack/linux-centos8-aarch64/gcc-8.2.1/readline-8.0-o72lkv3gc22wle6le25ijxakj522uf7i
[+] /home/xiaojun/spack/opt/spack/linux-centos8-aarch64/gcc-8.2.1/gdbm-1.18.1-bskc6oghdwgcgxhv6n233xqmfmgqvtqc
[+] /home/xiaojun/spack/opt/spack/linux-centos8-aarch64/gcc-8.2.1/perl-5.30.1-jjanez4jarykzlbl7o3dqo7ceehlloyz
[+] /home/xiaojun/spack/opt/spack/linux-centos8-aarch64/gcc-8.2.1/autoconf-2.69-r3la4hzi7zozk66oe2vf3velmbahmvcb
[+] /home/xiaojun/spack/opt/spack/linux-centos8-aarch64/gcc-8.2.1/automake-1.16.2-t46uzjcexxaistisynxkzg3ai7alsuby
[+] /home/xiaojun/spack/opt/spack/linux-centos8-aarch64/gcc-8.2.1/gmp-6.1.2-fidxg6nmtp5an5vohdtvfker5qtvqznk
==> 31754: Installing mpfr
==> Using cached archive: /home/xiaojun/spack/var/spack/cache/_source-cache/archive/c0/c05e3f02d09e0e9019384cdd58e0f19c64e6db1fd6f5ecf77b4b1c61ca253acc.tar.bz2
==> Fetching https://www.mpfr.org/mpfr-4.0.2/allpatches
################################################################################################################################################################################################# 100.0%
==> Error: ChecksumError: sha256 checksum failed for /tmp/root/spack-stage/spack-stage-68enuf8f/allpatches
    Expected f2d2a530acb5e70e1a9d5b80881dbb4a504d56535c4bc103d83e0bb630172029 but got 8f15fd27ab65341a60d724d594897d32f4597ddf642d0dc121995e2150181b0c

/home/xiaojun/spack/lib/spack/spack/package.py:1147, in do_fetch:
       1144
       1145        self.stage.cache_local()
       1146
  >>   1147        for patch in self.spec.patches:
       1148            patch.fetch()
       1149            if patch.stage:
       1150                patch.stage.cache_local()


==> Error: Failed to install mpfr due to ChildError: ChecksumError: sha256 checksum failed for /tmp/root/spack-stage/spack-stage-68enuf8f/allpatches
    Expected f2d2a530acb5e70e1a9d5b80881dbb4a504d56535c4bc103d83e0bb630172029 but got 8f15fd27ab65341a60d724d594897d32f4597ddf642d0dc121995e2150181b0c
/home/xiaojun/spack/lib/spack/spack/package.py:1147, in do_fetch:
       1144
       1145        self.stage.cache_local()
       1146
  >>   1147        for patch in self.spec.patches:
       1148            patch.fetch()
       1149            if patch.stage:
       1150                patch.stage.cache_local()

Traceback (most recent call last):
  File "/home/xiaojun/spack/lib/spack/spack/build_environment.py", line 802, in child_process
    return_value = function()
  File "/home/xiaojun/spack/lib/spack/spack/installer.py", line 1046, in build_process
    pkg.do_patch()
  File "/home/xiaojun/spack/lib/spack/spack/package.py", line 1178, in do_patch
    self.do_stage()
  File "/home/xiaojun/spack/lib/spack/spack/package.py", line 1163, in do_stage
    self.do_fetch(mirror_only)
  File "/home/xiaojun/spack/lib/spack/spack/package.py", line 1148, in do_fetch
    patch.fetch()
  File "/home/xiaojun/spack/lib/spack/spack/patch.py", line 194, in fetch
    self.stage.check()
  File "/home/xiaojun/spack/lib/spack/spack/stage.py", line 490, in check
    self.fetcher.check()
  File "/home/xiaojun/spack/lib/spack/spack/fetch_strategy.py", line 72, in wrapper
    return fun(self, *args, **kwargs)
  File "/home/xiaojun/spack/lib/spack/spack/fetch_strategy.py", line 504, in check
    "Expected %s but got %s" % (self.digest, checker.sum))
spack.fetch_strategy.ChecksumError: sha256 checksum failed for /tmp/root/spack-stage/spack-stage-68enuf8f/allpatches
    Expected f2d2a530acb5e70e1a9d5b80881dbb4a504d56535c4bc103d83e0bb630172029 but got 8f15fd27ab65341a60d724d594897d32f4597ddf642d0dc121995e2150181b0c
```

**Comment by darmac** [comment]
And I also found the patch for 4.0.2 will cause build-error.
If I remove the 4.0.2 patch, then it can build pass.

**Comment by adamjstewart** [comment]
Fixed by #15783
